### DISCUSSION TITLE
Is yarn pnp suitable for libraries too?

### DISCUSSION THREAD

**Discussion by abiriadev** [yarnpkg/berry#6288] [Q&A] [answered]
If I understood correctly, yarn pnp stores dependencies to git history. So if I'm writing a library, is it still worth using yarn pnp?

Does using pnp make fetching my library any slower or inefficient? Should I ignore `.yarn` from `npmignore`? What if I'm using my library as a [git dependency](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#git-urls-as-dependencies)? Have you used yarn pnp in your library? Then what was your experience?

Unfortunately, I couldn't find a guide or best practices from [the doc](https://yarnpkg.com/) on what should I do especially when I am writing a library. I appreciate all your answers and experiences!

**Comment by arcanis** [comment, marked as answer]
>  yarn pnp stores dependencies to git history

Not necessarily (starting from 4.0, PnP doesn't imply checking-in the dependencies by default). So there are multiple angles to this question:

**Should I use Yarn PnP when working on a library?**

To that I'd say yes, as PnP ensures you don't depend on packages without listing them in your dependencies. As a result your library is safer for your users, as there are less chances you'd accidentally publish something broken.

**Should I check-in my dependencies when working on a library?**

To that I'd say "probably not", although it depends on circumstances and your personal taste - checking in the dependencies has no impact once the package is published (it's only used in development), so it has neither pros nor cons from this perspective.

Checking-in dependencies is usually recommended in enterprise settings where you really don't want to access the network from your containers (be it because you're in an airgapped environment, or because you're worried of the registry's transient failures). For open-source libraries it's less of a concern, although it can be nice to avoid having to run `yarn install` when switching branches.

**Reply by abiriadev** [reply]
> starting from 4.0, PnP doesn't imply checking-in the dependencies by default

Oh, that's probably what I missed! Thank you for letting me know that!

> checking in the dependencies has no impact once the package is published

Now I understand this clearly. Regardless of whether dependencies are checked in, the published package will follow npm's dependency resolution process.

Thank you for your kind answer!
